### PR TITLE
Pq further rescoring optimization clean

### PR DESCRIPTION
**Pull request by abdelr**
### What's being changed:

When using PQ, we need to rescore vectors at some point to obtain good recall again. We have tried different policies and the best is to rescore at the very end, right before calculating the final k elements to retrieve.

Now, rescoring means we need to rebuild our current results list. With the current priority queue, this means popping all elements, calculating the distances again, and populating the queue. Popping elements is expensive since it reorders the elements again. On top of it, we no longer need the ef elements but only k elements. 

This PR introduces better memory allocation (by re-using existing allocated chunks), reseting capacity of priority queue when rescoring and checking before inserting a new rescored element if it could make it to the list. 

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [X] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [X] Performance tests have been run or not necessary.

### CODE REVIEW

**Comment by abdelr** [comment]
Quick comparison of performance

File: adapters/repos/db/vector/hnsw/priorityqueue/queue.go [hunk 1]
```diff
@@ -95,6 +95,12 @@ func (l *Queue) Pop() Item {
 	return out
 }
 
+func (l *Queue) SubstitudeTop(id uint64, distance float32) {
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/priorityqueue/queue.go:98]: Do you mean `SubstituteTop`? Maybe `ReplaceTop` is a better name?

File: adapters/repos/db/vector/ssdhelpers/product_quantization.go [hunk 2]
```diff
@@ -358,13 +359,20 @@ func (pq *ProductQuantizer) Fit(data [][]float32) {
 }
 
 func (pq *ProductQuantizer) Encode(vec []float32) []byte {
-	codes := make([]byte, pq.m)
+	codes := make([]byte, pq.m+4)
```
Comment (trengrj) [inline adapters/repos/db/vector/ssdhelpers/product_quantization.go:362]: Should these Distortion changes be in this commit?

File: adapters/repos/db/vector/hnsw/priorityqueue/queue.go (hunk 1 above)
Comment (abdelr) [reply adapters/repos/db/vector/hnsw/priorityqueue/queue.go]: Done

File: adapters/repos/db/vector/ssdhelpers/product_quantization.go (hunk 2 above)
Comment (abdelr) [reply adapters/repos/db/vector/ssdhelpers/product_quantization.go]: Definitely not... Removed already.

**Comment by abdelr** [comment]
Do not merge... there is a potential bug

**Comment by abdelr** [comment]
Closed in favour of https://github.com/weaviate/weaviate/pull/3363

**Pull request closed without merging**